### PR TITLE
[Test Framework] Integrate apps with Galley

### LINKED ISSUE #13090: Test Framework Native Apps Fail with Galley
```golang
g := galley.NewOrFail(t, ctx, galley.Config{})
pilotInstance := pilot.NewOrFail(t, ctx, pilot.Config{Galley: g})
instance := apps.NewOrFail(ctx, t, apps.Config{
    Pilot: pilotInstance,
})
```
Try to use the new framework, with native env. This is failing right now. if I set `Galley: nil` everything works fine.
Error is: `failed to configure apps: service a missing config for cluster outbound|34743||b.istio-system.svc.local. Dumping Envoy`

### PR DESCRIPTION
**Pull request by nmittler**
The most recent refactoring broke the apps component when Pilot is being used with Galley. The apps register their services with the ServiceManager directly. When Pilot is configured with Galley, however, it doesn't use the ServiceManager, which means that the app services are never properly registered with Pilot.

- Changed the Pilot and Apps component to require Galley to be configured, to avoid confusion.

- Removed the ServiceManager altogether - Galley is used for service registration.

Fixes #13090

### CODE REVIEW

**Comment by nmittler** [comment]
@howardjohn @smawson PTAL

I've added a hold for now. I've just pushed a major change to refactor the echo component so that it can be a complete replacement for the apps component. I've added full capabilities for both native and kube.  I've tested native, but will work tomorrow on swapping over the existing apps tests to use echo instead.

**Comment by nmittler** [comment]
@smawson 

> This PR is rather big, is it possible to split into more manageable pieces? Can we get here more incrementally?

Yes, sorry ... I've reverted the changes to echo and moved them over to #13175.

> - It feels like we're doing a lot of hackery around how we create the sidecars, I'd like us to figure out how to clean that up and have a roadmap for making this all much simpler. That may require work on our actual components to make them run the way we want, but we should do that anyway :)
> - I initially had planned on having a separate "sidecar" or "envoy" component rather than having apps take care of it, and using as much of the real code as possible. We should discuss how we want that to look in the long term at some point (I only have half-baked opinions but something doesn't quite feel right here).

Absolutely agree. This PR is really just taking baby steps to:

- Fix a bug in apps that was introduced with the new framework changes
- Merge 2 components that are really the same thing (apps and echo).

I've raised #13177 to track progress toward addressing the hackery. This is very important.

File: pkg/test/framework/components/apps/agent/bootstrap.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+      {{- end }}
+  {{- end -}}
+`
+)
+
+var (
+	envoyBootstrapTemplate = getEnvoyBootstrapTemplate()
+)
+
+func getEnvoyBootstrapTemplate() *template.Template {
```
[the remark this answers is not in the document]
Comment (nmittler) [reply pkg/test/framework/components/apps/agent/bootstrap.go]: Agreed. I've mentioned that in #13177

File: pkg/test/framework/components/echo/native/discoveryfilter.go [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+				// Exclude the iptables-mapped listener from the Envoy configuration. It's hard-coded to port 15001,
+				// which will likely fail to be bound.
+				continue
+			}
+
+			inbound, err := isInboundListener(l)
+			if err != nil {
+				return nil, err
+			}
+			if inbound {
```
[the remark this answers is not in the document]
Comment (nmittler) [reply pkg/test/framework/components/echo/native/discoveryfilter.go]: Understood. Do you think this should be addressed now, or can it wait for the resolutoion of #13177?

File: pkg/test/framework/components/apps/agent/serviceentry.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	}
+	return tmpl
+}
+
+func createServiceEntryYAML(a *Instance) ([]byte, error) {
+	var filled bytes.Buffer
+	w := bufio.NewWriter(&filled)
+	if err := serviceEntryTemplate.Execute(w, map[string]interface{}{
+		"ServiceName": a.cfg.ServiceName,
+		"Version":     a.cfg.Version,
```
Comment (howardjohn) [inline pkg/test/framework/components/apps/agent/serviceentry.go]: Why do we have version? Seems like we just set it as a label on the ServiceEntry, and we never have different versions of the same service so I don't see why it is useful

File: tests/integration/security/healthcheck/mtls_healthcheck_test.go [hunk 4]
```diff
@@ -74,8 +74,8 @@ spec:
 	if err != nil {
 		t.Error(err)
 	}
-	pilot := pilot.NewOrFail(t, ctx, pilot.Config{})
-	aps := apps.NewOrFail(ctx, t, apps.Config{Pilot: pilot, AppParams: []apps.AppParam{
+	p := pilot.NewOrFail(t, ctx, pilot.Config{})
+	aps := apps.NewOrFail(t, ctx, apps.Config{Pilot: p, AppParams: []apps.AppParam{
```
[the remark this answers is not in the document]
Comment (nmittler) [reply tests/integration/security/healthcheck/mtls_healthcheck_test.go]: Both now call validate. The namespace should have been optional ... we now use a default namespace if not provided.

File: pkg/test/framework/components/apps/native_test.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/apps"
+	"istio.io/istio/pkg/test/framework/components/environment"
+	"istio.io/istio/pkg/test/framework/components/galley"
+	"istio.io/istio/pkg/test/framework/components/namespace"
+	"istio.io/istio/pkg/test/framework/components/pilot"
+)
+
+func TestNative(t *testing.T) {
+	ctx := framework.NewContext(t)
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/test/framework/components/apps/native_test.go]: Looking at the code, looks like it just does some extra logging and ensures you close the context - not a big difference. Doesn't seem like a big deal, we may just want to change all tests to the same in a later PR for consistency

File: pkg/test/framework/components/environment/native/native.go [hunk 6]
```diff
@@ -35,19 +44,29 @@ type Environment struct {
...
+
 	e := &Environment{
-		ctx:            ctx,
-		Mesh:           &mesh,
-		ServiceManager: service.NewManager(),
+		ctx:             ctx,
+		Mesh:            &mesh,
+		SystemNamespace: systemNamespace,
+		Domain:          domain,
+		PortManager:     portMgr,
```
Comment (howardjohn) [inline pkg/test/framework/components/environment/native/native.go]: Looks like portmanager in Environment?

File: pkg/test/framework/components/apps/native.go [hunk 7]
```diff
@@ -101,37 +106,46 @@ func newNative(ctx resource.Context, cfg Config) (Instance, error) {
...
-		cfg.discoveryAddress = c.discoveryAddress
-		cfg.serviceManager = c.serviceManager
+	domain := ctx.Environment().(*native.Environment).Domain
+	tmpDir, err := ctx.CreateTmpDirectory("apps")
+	if err != nil {
+		return nil, err
+	}
 
-		app, err := newNativeApp(cfg)
+	portManager, err := reserveport.NewPortManager()
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/test/framework/components/apps/native.go]: https://github.com/istio/istio/pull/13115/files#diff-2368df1f890905899c2a72cb05f0afb2R69? Not sure if that links right, but it seems to be in this one?

[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/test/framework/components/apps/native.go]: If its cleaned up in the next one that is fine too

**Comment by nmittler** [comment]
@howardjohn I had to rebase ... PTAL

**Comment by howardjohn** [comment]
/lgtm
Remove the hold label if you want this in

**Pull request merged**